# Running Python in a Jupyter Notebook

The document you're currently viewing is a "Jupyter Notebook". Jupyter notebooks give you an interactive interface to different programming languages. In this case Python.

In this introduction you will learn some basics of Python by running (and sometimes editing) cells with Python code (the blueish "boxes" below). When a cell is run, all the Python code in that cell is executed, and you do this by selecting the cell and pressing Shift+Enter (alternatively, click the "Run" button above).

This introduction will mostly just ask you to run a bunch of code cells to see what happens, but sometimes you will also be asked to edit the contents of the cell. Don't feel limited by this however. *Every* cell is editable and interactive, so feel free to change things just to see what happens. Changes you make are autosaved, but should you mess things up, you can always undo your changes (select the cell and Ctrl-Z) or in the worst case, download the notebook again.

Note that *every* cell is editable. Even the instructions you are reading now. It might happen that you by mistake start editing the instructions (happens if you double click the text) in which case the formatting will look funny. To exit the edit mode, just run the cell (Shift+Enter) and it should look right again.

# First steps, Python as a calculator

Although not really the point, one thing you can do In the Python is simple calculations. Just enter an arithmetic expression in a cell, run it by pressing Shift-Enter (or the Run button above), and Python will calculate it if it can. The cells below already have some arithmetic expressions in them. Select each cell, run it and see what happens!

In [1]:
1+1

2

In [2]:
1+3*2**3

25

The second example might look a bit funny. The symbol * is (unsurprisingly)
used to denote multiplication, while ** denotes exponentiation. In other words, the Python expression $\texttt{1+3*2**3}$ corresponds to the mathematical expression $1+3\cdot 2^3$ which of course evaluates to 25.

If you want to, go ahead and change the content of the cells above, run them again and see what happens. Feel free to experiment!

A point to note that is important for AE 1, is that Python can also handle complex numbers. Just use `j` for the imaginary unit. Try running the cells below and see what happens!

In [3]:
1j**2

(-1+0j)

In [4]:
(1+2j)*1j

(-2+1j)

In the first example above, we calculated $i^2$ which is $−1$. To enter $i$ we had
to type $\texttt{1j}$. In the second example we calculated $(1 + 2i)\cdot i$ which evaluates as

$(1 + 2i)\cdot i = i + 2i^2 = −2 + i.$

If you are used to other mathematical or statistical software such as Matlab
or R, you might find that Python appears (at first glance) to have very limited mathematical capabilities. For example, in Matlab you could type `sin(1)` and
immediately get an approximate value of $\sin 1$, but if you try the same in Python it doesn’t seem to work (run the cell below and you'll get an error message). Don’t worry though. There’s a huge library of mathematical functions (including of course all the standard trigonometric functions) available in Python. We just have to learn how to access them, and we will do so in AE2.

In [5]:
sin(1)

NameError: name 'sin' is not defined

# Using variables

You can use the `=` symbol to assign values to variable names and use them in
your calculations. Run the cell below and see what happens!

In [6]:
a=3
b=2
a+b

5

It is important to note that in Python, the `=` symbol should be thought of as an *assignment* operator and not as a mathematical equal sign. What this means is that when Python encounters the `=` symbol, it takes what’s on the right side of the symbol, evaluates it, and assigns it to the variable name on its left side. 

The following manipulations illustrate how it works (and how it is different from mathematical equality). Run the cells below in the order they are written (start with the topmost and proceed downwards), but before you do so, try to guess what the output should be!

In [7]:
k=1
k=k+1
k  #This line generates the current value of k as output

2

In [8]:
k=k+2
k

4

In [9]:
k=(k+5)/3
k

3.0

Is each step above, Python takes the current value of the right hand side and assigns it to the left hand side. First we assign the value $1$ to $k$. Then we take the current value $(1+1)$ of $k+1$ and assign it to $k$ so it becomes $2$. 

In the next step we take the current value $(2+2)$ of $k+2$ and assign it to $k$ which is now $4$.

Of course, if you run any of the cells more than once, or in a different order, different things will happen. Feel free to experiment and see what happens!

The examples above illustrates that the `=` symbol should not be thought of as mathematical equality (the mathematical equation $k = k + 1$ has no solutions). Instead we should think of it as *assigning* the value on the right to the variable on the left.

# Logical expressions

While programming in Python, we will often want to test whether different conditions are true or false. In particular, we would like to check whether certain equalities or inequalities are satisfied. In Python, the comparison operators `< > <= >= ==` correspond to the mathematical relations $< \ > \ \leq \ \geq \ =$. Note that
since in Python `=` is used to assign values, we must use a different symbol `==` to represent equality. 

Python will test the condition you give, return `True` if it
is true and return `False` if it is false. Try running each cell below and see what happens (but please try to guess what the result should be before you run it).

In [10]:
2+3 == 7-2

True

In [11]:
2+3 == 7

False

In [12]:
2+3 < 7

True

In [13]:
2+3 >= 7

False

We can combine two logical expressions into one using the `and`, and `or` logical
expressions. The expression `P and Q` returns `True` if both `P` and `Q` are true, and
returns `False` otherwise. Run the cells below to see how it works. And, as usual, feel free to experiment.

In [14]:
2<3 and 4<5

True

In [15]:
2<3 and 4==5

False

In [16]:
2>3 and 4<5

False

In [17]:
2>3 and 4==5

False

The expression `P or Q` returns `True` if `P` or `Q` (or both) are true. It returns `False` only if `P` and `Q` are both false. Try the following!

In [18]:
2<3 or 4<5

True

In [19]:
2<3 or 4==5

True

In [20]:
2>3 or 4<5

True

In [21]:
2>3 or 4==5

False

The `not` operator negates the expression that follows. E.g.:

In [22]:
not False

True

In [23]:
not True

False

In [24]:
not 2+3 == 7-2

False

In [25]:
not 2+3 == 7

True

# Functions in Python
In Python, a function is essentially a block of instructions that carry out a
specific task. To perform the task, the function usually (but not necessarily) needs some
input (one or several "arguments") and it usualy (but not necessarily) returns some kind of result.

### Using built in functions
A multitude of functions can be accessed directly in Python. For example, if we
want to extract the smallest out of a collection of values, we can use the built in
`min` function. Run the cells below to see how it works!

In [26]:
min(2,3)

2

In [27]:
a=5
b=7
min(a,b)

5

In [28]:
min(1,-3,a,b)

-3

As illustrated, the `min` function takes one or several values as input and returns
the smallest of those values. In the various Application Exercises that follow,
we will encounter many other built in functions.

### Making your own functions, simple example

If we need a function that is not already built in we can, with a little effort,
make it ourselves.

Let's say that we want to be able to type `square(x)` in order to calculate $x^2$ (this is of course silly since we can easily do this by typing $\texttt{x**2}$, but let's consider this as an easy example). There is no built in function called $\texttt{square}$ so if you run the cell below (try it!) you'll get an error message.

In [31]:
square(3)

9

Ok, that (of course) didn't work, but now let's do the following.

Run the cell below!

In [30]:
def square(x) :
    """Returns the square of its argument"""
    y=x**2
    return(y)

Ok, after running it, nothing happened (it seems)!

But... now try running the previous cell (the one where with `square(3)` and which returned an error message) again.

If everything works, this cell now returns `9`. If you want, go ahead and try other values!

### Making your own functions, the general structure
The general syntax for creating your own function in Python is the following.

    def function_name(parameters) :
        """docstring"""
        instructions```    

To clarify
1. The function header (first row) of a function declaration must start with
the keyword def, followed by the name of your function and brackets
containing any parameter names for arguments passed to the function (if the function takes more than one parameter - these are separated by commas, if it takes no parameters - just leave the brackets empty),
and finally end with a colon. In our example (the cell above) we defined a function
called square taking one argument that is assigned the parameter name
`x`.
2. Following the function header comes an optional docstring. The docstring
is a text within triple quotation marks that describes in normal language
what the function does. You don’t have to provide a docstring with your
function, but it is good practice to do so.
3. After the docstring comes one or several rows of code (instructions) that
perform the actual task of the function. These rows must have the same
indentation level, conventionally 4 spaces, relative to the function header.
All lines with the same or larger indentation level are considered to be
part of the function.
4. Typically, a function will return something as well. In our example (the cell above) the function returns the value of the variable `y` which has been assigned the value of `x**2`. In other words, this function returns the value of `x**2`.

After a function has been defined, you can use it in Python like any other. For example, you can try running the cells below and see what happens.

In [32]:
square(-2)+square(3)


13

In [33]:
x = square(3)
y = square(4)
(x+y)**0.5

5.0

### Exercise
In the cell below, create a function called `hypothenuse` that takes two arguments and returns the square root of the sum of their squares. I have already created a function header for you, so all you have to do is write the rest of the function. 

In the cell below, is the function header. Edit the cell and fill in the rest of the function. Then run the cell.

In [34]:
def hypothenuse(x,y) :
    return (x**2 + y**2)**0.5
    

If this is the first time you try to write a function in Python, you will probably make some mistake(s) and get an error message. Just keep trying until running the cell above gives no error message. 

*Hint 1*: You can calculate square roots by taking a suitable power.

*Hint 2*: One more line of code is enough. Just `return` the appropriate value. 

*Hint 3*: Python uses indentation of code to tell where the function code begins and ends (and as we will se later, other chunks of code called code blocks). In this example, all lines of code following the function header should be indented by the same number of blank spaces (typically 4). This will probably be done automatically once you start editing the cell above, but if you mess it up, you will get an error message.

If you really can't get it to work, you'll find a solution at the end of this notebook.

Finally *test* your function by running the two cells below!

In [35]:
hypothenuse(3,4)

5.0

In [36]:
hypothenuse(4,5)

6.4031242374328485

If your function works properly, the first cell should return `5.0` while the second should return `6.4031242374328485`. If this doesn't work, try to edit the code for the `hypothenuse` function until it works.

## While loops
Suppose we don't know that

$$1+2+\ldots+n=\frac{n(n+1)}{2}$$

In that case we could write a Python function that calculates the sum above in the straightforward manner, just by performing the necessary additions. This function below, using a `while` loop (we will explain later how it works), does the trick (If you have some programming experience, you’ll probably realize that a `for` loop would be easier to use in this case, but this is just an example for illustration purpose):

In [37]:
def addup(n) :
    """ Adds all integers starting with 1 and ending with n """
    k=1
    s=0
    while k <= n :
        s=s+k
        k=k+1
    return s

Run the cell above so the function gets defined. Then try it out by running the cells below.

In [38]:
addup(1)

1

In [39]:
addup(3)

6

In [40]:
addup(100)

5050

Let’s take a closer look at the `addup` function and how it works:

The code was:

    def addup(n) :
        """ Adds all integers starting with 1 and ending with n """
        k=1
        s=0
        while k <= n :
            s=s+k
            k=k+1
        return s

And this is how it works:
1. The function header defines a function called `addup` taking one argument
`n`.
2. In the two first lines (after the docstring) we set the variables `k=1` and
`s=0`. In this example `k` will act as a counter and `s` will be the calculated
sum (the value that the function eventually returns).
3. The next code block, starting with `while k<=n :` followed by two indented lines, is a `while`-loop. It keeps executing the indented lines, repeatedly over and over again, as long as the condition `k<=n` is satisfied.

*IMPORTANT* Note the indentation of the code. The two lines after the `while` statement are indented relative to the rest of the code. The indented code (and *only* the indented code) gets executed repeatedly as long as the `while`-condition is satisfied. When (eventually) the condition is no longer satisfied, the Pyhon proceeds with the code *after* the indented lines. 
4. Finally (*after* the `while`-loop has finished), the function returns the value of `s`.

To understand in detail how it works, suppose we call the function with `n=3` like we did in one of the cells above. Let us step by step go through how the steps of the functions were executed in that case:

1. The first line of code: `k=1`. The counter `k` is set to `1`.
2. `s=0`. The sum `s` is set to be `0`.
3. `while k<=n :`. Python checks the condition. Since currently `k=1` and `n=3`,
the condition is `True` and Python proceeds to execute the following
indented lines.
4. `s=s+k`. We update the value of the sum `s` to be the current value of `s+k`, which is
0 + 1 = 1.
5. `k=k+1`. We update the value of our counter `k` by adding `1` to its current value, so now
`k` has the new value `2`.
6. After executing the last indented line, Python jumps back to the start of
the loop.
7. `while k<=n :`. Python checks the condition again. Since currently `k=2` and `n=3`,
the condition is `True` and Python proceeds to execute the commands
in the loop.
8. `s=s+k`. We update the value of `s` to be the current value of `s+k`, which is
1 + 2 = 3.
9. `k=k+1`. We update the value of `k` by adding `1`, so now `k` has the value 3.
10. Python jumps back to the start of the loop.
11. `while k<=n :`. Python checks the condition. Since currently `k=3` and
`n=3`, the condition is `True` and the commands in the loop are executed
again.
12. `s=s+k`. We update the value of `s` to be the current value of `s+k`, which is
3 + 3 = 6.
13. `k=k+1`. We update the value of `k` by adding `1`, so now `k=4`.
14. Python jumps back to the start of the loop.
15. `while k<=n :`. Python checks the condition. Since currently `k=4` and
`n=3`, the condition is `False` and Python skips to the next line (if
any) after the last indented line.
16. The next line after the loop returns the current value of `s` which is `6`.

If the steps above seem overwhelming to process, just note that we start by
setting s=0, then successively we add 1, 2 and 3 to s, so when the process exits,
we have s=6.

### Example

Let's do one more example. This one is a little bit more complicated, but if you can undestand how it works, you will be ready for Application Exercise 1.

Let
$$s_n=4\cdot\sum_{k=0}^n \frac{(-1)^k}{2k+1}=4\cdot\left(1-\frac{1}{3}+\frac{1}{5}+\ldots +\frac{(-1)^n}{2n+1}\right).$$
It is possible to show (you will learn about this later) that
$$\lim_{n\to\infty}s_n=\pi,$$
which gives us a method (although, rather inefficient) to approximate $\pi$: 
#### Calculate $s_n$ for a large enough value of $n$ and we should get a good approximation of $\pi$.
Even better, noting that
$$s_n=\sum_{k=0}^n (-1)^k a_k,\quad\text{where }a_k=\frac{4}{2k+1},$$
there's a result that tells us that
$$|s_n-\pi|<a_{n+1},$$
in other words, if we approximate $\pi$ with $s_n$, the absolute error is smaller than the modulus of the first omitted term, i.e. smaller than 
$$a_{n+1}=\frac{4}{2n+3}.$$
Let's implement this in Python. Let's create a function which returns an approximate value of $\pi$ with absolute error smaller than 0.0005.

Here is such a function (run the cell so that the function gets defined):

In [43]:
def findpi(error) :
    """Return an approximate value of pi"""
    k=0                         # Start with k=0.
    s=4                         # Let s be the value of the first term.
    while 4/(2*k+3) >= error :  # Loop as long as the next term has modulus at least equal to the allowed error.
        k=k+1                   # Step up the counter k
        s=s+(-1)**k*4/(2*k+1)   # Add the k:th term to s
    return s                    # After the loop is done, return the value of s

Note that we can add comments to our code using the `#`-symbol. Everything after the `#`-symbol is ignored by Python so you can write whatever you want (typically clarifying comments), without affecting your code.

Now, let's call the function (run the following cell). Note that the function takes no argument, so we just call it as `findpi()` with nothing between the brackets.

In [45]:
findpi(0.0005)

3.1413426535937043

Looks ok. The first three decimals are correct, as they should be. The remaining decimals are wrong, but the error is within the allowed limit.

### Exercise

Note that the function `findpi()` takes no argument. Your last exercise before we do Application Exercise 1, is to modify the function above (edit the cell where the function is defined and then re-run that cell so your edits take effect). After you are done editing, the function should now take one argument (you can call it what you want) and it should return an approximation of $\pi$ with error smaller than that argument.

*Hint*: You only have to change the code in two places to make this work.

Next let's try your code by running the two cells below. 

In [46]:
findpi(0.000005)

3.141590153589744

In [47]:
findpi(0.0000005)

3.141592403589655

If your code is correct, the first cell should produce the output `3.141590153589744`, while the second should give you `3.141592403589655`.

If that works - Great! If not, try to find out what went wrong.

A final comment: If you try calling the function with an even smaller argument, you'll probably find that it takes very long for the command to execute (the cell is marked with a star as long as the code is running). This is because the method we are using is actually very inefficient, so calculating $\pi$ this way requires a lot of work for your computer.

# You are now ready for AE1

---

# Answers to exercises:

#### Exercise 1
For the first exercise (the `hypothenuse` function), the following code will do the job:

    def hypothenuse(x,y) :
        return (x**2+y**2)**0.5

    
#### Exercise 2
The following function takes an error bound as argument and returns an approximation of $\pi$ within that error:

    def findpi(error) :
        """Return an approximate value of pi with error smaller than the argument"""
        k=0                         # Start with k=0.
        s=4                         # Let s be the value of the first term.
        while 4/(2*k+3) >= error :  # Loop as long as neded.
            k=k+1                   # Step up the counter k
            s=s+(-1)**k*4/(2*k+1)   # Add the k:th term to s
        return s                    # After the loop is done, return the value of s